File: IAIP.Testx/ControlTests/AirsNumberTextBoxTests.cs
using Iaip;
using Iaip.DAL;
using IAIP.Testx.ExtensionMethods;
using Xunit;

namespace IAIP.Testx.ControlTests
{
    public class AirsNumberTextBoxTests
    {
        private AirsNumberTextBox airs;

        public AirsNumberTextBoxTests()
        {
            airs = new AirsNumberTextBox();
        }

        [Fact]
        private void InitialValues()
        {
            Assert.Null(airs.AirsNumber);
            Assert.Equal(9, airs.MaxLength);
            Assert.Equal("000-00000", airs.Cue);
            Assert.False(airs.FacilityMustExist);
            Assert.Equal(AirsNumberValidationResult.Empty, airs.ValidationStatus);
        }

        [Theory]
        [InlineData("111")]
        [InlineData("ABC")]
        [InlineData("0010001")]
        [InlineData("001-0001")]
        private void InvalidFormat(string input)
        {
            airs.Text = input;
            bool result = airs.Validate();

            Assert.True(result);
            Assert.Equal(AirsNumberValidationResult.InvalidFormat, airs.ValidationStatus);
            Assert.Equal(input, airs.Text);
        }
        
        [Theory]
        [InlineData("00100001")]
        [InlineData("001-00001")]
        [InlineData("99999999")]
        [InlineData("999-99999")]
        private void ValidFormat(string input)
        {
            airs.Text = input;
            bool result = airs.Validate();

            Assert.True(result);
            Assert.Equal(AirsNumberValidationResult.Valid, airs.ValidationStatus);
            Assert.Equal(string.Concat(input.Substring(0, 3), "-", input.Substring(input.Length - 5)), airs.Text);
        }

        [Theory]
        [InlineData(null)]
        [InlineData("")]
        private void EmptyInput(string input)
        {
            airs.Text = input;
            bool result = airs.Validate();

            Assert.True(result);
            Assert.Equal(AirsNumberValidationResult.Empty, airs.ValidationStatus);
            Assert.Equal(string.Empty, airs.Text);
        }
        
        [Fact]
        private void ValidationStatusChanged()
        {
            airs.ValidationStatusChanged += Airs_ValidationStatusChanged;

            Assert.Null(airs.Tag);

            airs.Text = "001-00001";
            airs.Validate();

            Assert.NotNull(airs.Tag);
            Assert.True((bool) airs.Tag);
        }

        private void Airs_ValidationStatusChanged(object sender, System.EventArgs e)
        {
            airs.Tag = true;
        }
    }
}

File: IAIP.Test/ControlTests/AirsNumberTextBoxTests.vb
Imports IAIP.DAL
Imports Xunit

Public Class AirsNumberTextBoxTests

    <Fact>
    Private Sub InitialValues()
        Using airsTextBox As New AirsNumberTextBox()
            Assert.Null(airsTextBox.AirsNumber)
            Assert.Equal(9, airsTextBox.MaxLength)
            Assert.Equal("000-00000", airsTextBox.Cue)
            Assert.False(airsTextBox.FacilityMustExist)
            Assert.Equal(AirsNumberValidationResult.Empty, airsTextBox.ValidationStatus)
        End Using
    End Sub

    <Fact>
    Private Sub InvalidFormat()
        Dim input As String = "invalid"

        Using airsTextBox As New AirsNumberTextBox With {.Text = input}
            Dim result As Boolean = airsTextBox.Validate()

            Assert.True(result)
            Assert.Equal(AirsNumberValidationResult.InvalidFormat, airsTextBox.ValidationStatus)
            Assert.Equal(input, airsTextBox.Text)
        End Using
    End Sub

    <Fact>
    Private Sub ValidInput()
        Dim input As String = "00199999"

        Using airsTextBox As New AirsNumberTextBox With {.Text = input}
            Dim result As Boolean = airsTextBox.Validate()

            Assert.True(result)
            Assert.Equal(AirsNumberValidationResult.Valid, airsTextBox.ValidationStatus)
            Assert.Equal(String.Concat(input.Substring(0, 3), "-", input.Substring(input.Length - 5)), airsTextBox.Text)
        End Using
    End Sub

    <Theory>
    <InlineData(Nothing)>
    <InlineData("")>
    Private Sub EmptyInput(input As String)
        Using airsTextBox As New AirsNumberTextBox With {.Text = input}
            Dim result As Boolean = airsTextBox.Validate()

            Assert.True(result)
            Assert.Equal(AirsNumberValidationResult.Empty, airsTextBox.ValidationStatus)
            Assert.Equal(String.Empty, airsTextBox.Text)
        End Using
    End Sub

    <Fact>
    Private Sub ValidationStatusChanged()
        Using airsTextBox As New AirsNumberTextBox()
            AddHandler airsTextBox.ValidationStatusChanged, AddressOf Airs_ValidationStatusChanged

            Assert.Null(airsTextBox.Tag)

            airsTextBox.Text = "001-00001"
            airsTextBox.Validate()

            Assert.NotNull(airsTextBox.Tag)
            Assert.True(CBool(airsTextBox.Tag))
        End Using
    End Sub

    Private Sub Airs_ValidationStatusChanged(sender As Object, e As EventArgs)
        Dim airsTextBox As AirsNumberTextBox = CType(sender, AirsNumberTextBox)
        airsTextBox.Tag = True
    End Sub

End Class
